Fail snap refresh/revert jobs if the `snap_update_test.py` script fails (bugfix)

pieqq opened the pull request:
<!--
Example Title: Fixed bugged behaviour of checkbox load config (Bugfix)

A Traceability Marker is required as a suffix in the PR title to help understand the impact of your change at a glance.

Pick one of the following:
- Infra: Your change only includes documentation, comments, github actions or metabox
- BugFix: Your change fixes a bug
- New: Your change is a new backward compatible feature, a new test/test plan/test inclusion
- Breaking: Your change breaks backward compatibility.
    - This includes any API change to checkbox-ng/checkbox-support
    - Changes to PXU grammar/field requirements
    - Breaking changes to dependencies in snaps (fwts upgrade for example)

If your change is to providers it can only be (Infra, BugFix or New).

If your change impacts the submission format in Checkbox test reports, ensure that `submission-schema/schema.json` is updated and relevant fields are documented.

Signed commits are required.
  - See CONTRIBUTING.md (https://github.com/canonical/checkbox/blob/main/CONTRIBUTING.md#signed-commits-required) for further instructions.
  - If you are posting your first pull request from a fork of the repository, a Checkbox maintainer (someone with contributor / maintainer / admin rights) will be required to enable CI checks in the repo to be executed.
    - This will be communicated with a comment to the PR of the form `/canonical/self-hosted-runners/run-workflows <SHA-for-HEAD-commit>`
-->

## Description

Adding `set -eo pipefail` in the snapd refresh/revert jobs that initiate
a reboot ensures that:

- any failure in the `snap_update_test.py` is not masked by the pipe
command (this is achieved by `set -o pipefail`)
- if an exception occurs while issuing a refresh or revert snap command, the script records the issue using a special Checkbox mechanism (`__result` file in the session sharing directory) and raise an exception. This will not prevent the job from rebooting, but when resuming, the job will be marked as failed.

Minor: increase timeout to 10 minutes to try to limit the number of issues on slow devices.


<!--
Describe your changes here:

- What's the problem solved (briefly, since the issue is where this is elaborated in more detail).
- Introduce your implementation approach in a way that helps reviewing it well.
- Alert the reviewer of any changes that involve data persistence: present examples of file format changes as part of the PR description (e.g. new fields of data stored in unit or submission output).
-->

## Resolved issues

<!--
Note the Jira and GitHub issue(s) resolved by this PR (`Fixes|Resolves ...`).
Make sure that the linked issue titles & descriptions are also up to date.
-->

- CHECKBOX-1671
- #1615 

## Documentation

<!--
Please make sure that...
- Documentation impacted by the changes is up to date (becomes so, remains so).
  - Documentation in the repository, including contribution guidelines.
  - Process documentation outside the repository.
- Tests are included for the changed functionality in this PR. If to be merged without tests, please elaborate why.
- When breaking changes and other key changes are introduced, the PR having been merged should be broadcast (in demo sessions, IM, Discourse) with relevant references to documentation. This is an opportunity to gather feedback and confirm that the changes and how they are documented are understood.
-->

## Tests

<!--
- How was this PR tested? Please provide steps to follow so that the reviewer(s) can test on their end.
- Please provide a list of what tests were run and on what platform/configuration.
- Remember to check the test coverage of your PR as described in CONTRIBUTING.md
-->

Hook25 reviewed (changes requested):
Please confirm that noreturn will actually terminate when a job fails

pieqq commented:
> Please confirm that noreturn will actually terminate when a job fails

Unfortunately you're right, a `noreturn` job gets stuck there.

Hook25 commented:
You can try to use `$PAINBOX_SESSION_SHARE/__result`, this will make a pointless `restart`, but it at least gives the right outcome

pieqq commented:
I've side-loaded the provider on the same device that exhibited the original problem, and the jobs seem to pass now:

```
(...)
----------------------------[ Running job 26 / 37 ]-----------------------------
--------------[ Refresh cascade-kernel snap to its base revision ]--------------
ID: com.canonical.certification::snapd/snap-refresh-kernel-cascade-kernel-to-base-rev
Category: Snapd
--------------------------------------------------------------------------------
Refreshing snap cascade-kernel from revision 203 to 155
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Copy snap "cascade-kernel" data
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Setup snap "cascade-kernel" (155) security profiles
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Setup snap "cascade-kernel" (155) security profiles
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Setup snap "cascade-kernel" (155) security profiles
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Wait) Make snap "cascade-kernel" (155) available to the system
Snap operation finished. See `snap change 1578` for more info
Waiting for reboot...
Connection lost!
[Errno 32] Broken pipe
Reconnecting ...
Reconnected (took: 387s)
----------------------------[ Running job 28 / 38 ]-----------------------------
-----[ Attach logs after refreshing cascade-kernel snap to base revision ]------
ID: com.canonical.certification::snapd/log-attach-after-snap-refresh-kernel-cascade-kernel-to-base-rev
Category: Snapd
--------------------------------------------------------------------------------
Refreshing snap cascade-kernel from revision 203 to 155
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Copy snap "cascade-kernel" data
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Setup snap "cascade-kernel" (155) security profiles
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Setup snap "cascade-kernel" (155) security profiles
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Doing) Setup snap "cascade-kernel" (155) security profiles
(info) (Doing) Monitoring snap "cascade-kernel" to determine whether extra refresh steps are required
(info) (Wait) Make snap "cascade-kernel" (155) available to the system
Snap operation finished. See `snap change 1578` for more info
Waiting for reboot...
--------------------------------------------------------------------------------
Outcome: job passed
----------------------------[ Running job 29 / 38 ]-----------------------------
---[ Verify cascade-kernel snap revision after refreshing to base revision ]----
ID: com.canonical.certification::snapd/snap-verify-after-refresh-kernel-cascade-kernel-to-base-rev
Category: Snapd
--------------------------------------------------------------------------------
Beginning verify...
cascade-kernel snap refresh complete
Checking refresh status for snap cascade-kernel...
PASS: current revision (155) matches the expected revision
--------------------------------------------------------------------------------
Outcome: job passed
----------------------------[ Running job 30 / 38 ]-----------------------------
-----[ Revert cascade-kernel snap from base revision to original revision ]-----
ID: com.canonical.certification::snapd/snap-revert-kernel-cascade-kernel-from-base-rev
Category: Snapd
--------------------------------------------------------------------------------
Reverting snap cascade-kernel from revision 155 to 203
(info) (Doing) Setup snap "cascade-kernel" (203) security profiles
(info) (Doing) Setup snap "cascade-kernel" (203) security profiles
(info) (Doing) Setup snap "cascade-kernel" (203) security profiles
(info) (Wait) Make snap "cascade-kernel" (203) available to the system
Snap operation finished. See `snap change 1579` for more info
Waiting for reboot...
Connection lost!
[Errno 104] Connection reset by peer
Reconnecting ...
Reconnected (took: 364s)
----------------------------[ Running job 31 / 38 ]-----------------------------
------[ Attach logs after reverting cascade-kernel snap to base revision ]------
ID: com.canonical.certification::snapd/log-attach-after-snap-revert-kernel-cascade-kernel-from-base-rev
Category: Snapd
--------------------------------------------------------------------------------
Reverting snap cascade-kernel from revision 155 to 203
(info) (Doing) Setup snap "cascade-kernel" (203) security profiles
(info) (Doing) Setup snap "cascade-kernel" (203) security profiles
(info) (Doing) Setup snap "cascade-kernel" (203) security profiles
(info) (Wait) Make snap "cascade-kernel" (203) available to the system
Snap operation finished. See `snap change 1579` for more info
Waiting for reboot...
--------------------------------------------------------------------------------
Outcome: job passed
----------------------------[ Running job 32 / 38 ]-----------------------------
---[ Verify cascade-kernel snap revision after reverting from base revision ]---
ID: com.canonical.certification::snapd/snap-verify-after-revert-kernel-cascade-kernel-from-base-rev
Category: Snapd
--------------------------------------------------------------------------------
Beginning verify...
cascade-kernel snap revert complete
Checking revert status for snap cascade-kernel...
PASS: current revision (203) matches the expected revision
--------------------------------------------------------------------------------
Outcome: job passed
(...)
```

pieqq commented:
I've modified the script to timeout much earlier (a few seconds instead of 10 minutes) just to trigger some exceptions. I noticed another exception needed to be caught in order to apply the trick using `__result` file, so I added a patch for this.

I ran the modified script on the same device, and when the session is auto-resumed, the kernel refresh job is automatically marked as failed as expected (this is an abstract from the raw session:

```
      "com.canonical.certification::snapd/snap-refresh-kernel-cascade-kernel-to-base-rev": [
        {
          "comments": "Doing",
          "execution_duration": 493.60218334198,
          "io_log": [],
          "outcome": "fail",
          "return_code": null
        }
      ],
```

Hook25 commented on providers/base/bin/snap_update_test.py:
```diff
@@ -148,11 +151,25 @@ def snap_refresh(self):
+            if checkbox_session_dir:
+                result = {
+                    "outcome": "fail",
+                    "comments": exc.message,
```
Consider changing this to:
```suggestion
                    "comments": "Marking the test as failed because it raised the following:" + str(exc),
```
